src/protobuf-net.BuildTools/Internal/CodeGenSemanticModelParser.cs

```diff
@@ -2,170 +2,88 @@
 using Microsoft.CodeAnalysis;
 using ProtoBuf.Reflection.Internal.CodeGen;
 using System;
-using System.Collections.Generic;
-using System.Collections.Immutable;
-using System.Diagnostics;
 using System.Linq;
-using System.Text;
-using System.Xml.Linq;
+using System.Collections.Generic;
 
-namespace ProtoBuf.BuildTools.Internal;
+namespace ProtoBuf.Internal;
 
-internal static class CodeGenSemanticModelParser
+internal static partial class CodeGenSemanticModelParser
 {
     [ThreadStatic]
-    private static Stack<ISymbol> s_SpareQueue;
+    private static Stack<(ISymbol compilerSymbol, object codeGenType)> s_SpareQueue;
+
+    private static Stack<(ISymbol compilerSymbol, object codeGenType)> GetStack() => s_SpareQueue ?? new();
 
-    private static Stack<ISymbol> GetStack() => s_SpareQueue ?? new();
-    private static void RecycleStack(Stack<ISymbol> value)
+    private static void RecycleStack(Stack<(ISymbol compilerSymbol, object codeGenType)> value)
     {
         if (value is not null && s_SpareQueue is null)
         {
             value.Clear();
             s_SpareQueue = value;
         }
     }
+
     internal static void Parse(CodeGenSet set, ISymbol symbol)
     {
         CodeGenFile? file = null;
+        object? currentCodeGenType = null;
+        
         var pending = GetStack(); // prevent deep thread-stack dive by using an explicit stack
-        pending.Push(symbol);
+        pending.Push((compilerSymbol: symbol, codeGenType: null)!);
```

Comment:
shouldn't this be `codeGenType: null!));` ? i.e. move the `!` ? or maybe just use `, object? codeGenType`? still looking into what this arg does, so take my comments with a pinch of salt!

Reply:
since I have refactored code to avoid stack approach - comment is outdated. I am sure you will find many troubles with new code thou!